pkg/controllers/traffictopology/adapter.go

```diff
@@ -0,0 +1,532 @@
+// Copyright 2023 The KusionStack Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package traffictopology
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"kusionstack.io/kube-utils/multicluster/clusterinfo"
+	rsFrameController "kusionstack.io/resourceconsist/pkg/frame/controller"
+	"kusionstack.io/rollout/apis/rollout/v1alpha1"
+	"kusionstack.io/rollout/pkg/utils"
+)
+
+type TPControllerAdapter struct {
+	client.Client
+}
+
+func NewTPControllerAdapter(c client.Client) *TPControllerAdapter {
+	return &TPControllerAdapter{
+		c,
+	}
+}
+
+var _ rsFrameController.ReconcileAdapter = &TPControllerAdapter{}
+var _ rsFrameController.ReconcileWatchOptions = &TPControllerAdapter{}
+var _ rsFrameController.StatusRecordOptions = &TPControllerAdapter{}
+var _ rsFrameController.MultiClusterOptions = &TPControllerAdapter{}
+
+func (t *TPControllerAdapter) GetControllerName() string {
+	return ControllerName
+}
+
+func (t *TPControllerAdapter) GetSelectedEmployeeNames(ctx context.Context, employer client.Object) ([]string, error) {
+	return nil, nil
+}
+
+func (t *TPControllerAdapter) GetExpectedEmployer(ctx context.Context, employer client.Object) ([]rsFrameController.IEmployer, error) {
+	var expected []rsFrameController.IEmployer
+
+	trafficTopology, ok := employer.(*v1alpha1.TrafficTopology)
+	if !ok {
+		return expected, fmt.Errorf("not type of TrafficTopology")
+	}
+
+	backendApiVersion := ""
+	backendKind := "Service"
+	if trafficTopology.Spec.Backend.APIVersion != nil {
+		backendApiVersion = *trafficTopology.Spec.Backend.APIVersion
+	}
+	if trafficTopology.Spec.Backend.Kind != nil {
+		backendKind = *trafficTopology.Spec.Backend.Kind
+	}
```

Comment:
use ptr.Deref to simplify the code

backendApiVersion := ptr.Deref(trafficTopology.Spec.Backend.APIVersion, "")
backendKind := ptr.Deref(trafficTopology.Spec.Backend.Kind, "Service")